File: C:C++/fib.c
#include <stdio.h>

int fibonacci(int num, int* memo){

    if (memo[num] == -1){
        memo[num] = fibonacci(num-1, memo) + fibonacci(num-2, memo);
    }

    return memo[num];
}

int main(){
    int N = 10;
    int cache[N];
    for(int i = 0; i < N; i++){
        cache[i] = -1;
    }

    for(int i = 0; i < N; i++){
        printf("%d", cache[i]);
    }    
    
    cache[0] = 1;
    cache[1] = 1;

    printf("\n%d\n", fibonacci(N-1, cache));
  
    return 0;
} 

File: algorithm/dynamic programming/fib.py
import fibonacci
from fibonacci import fibo

def fib(num:int, memo = {}):

    if num in memo:
        return memo[num]

    if num == 0:
        return 0
    if num == 1:
        return 1

    memo[num] = fib(num-2)+fib(num-1)

    return memo[num]

import unittest

class TestCalculations(unittest.TestCase):

    def test_sum(self):
        for i in range(100):
            self.assertEqual(fibonacci.fibo(i), fib(i), "Something went wrong.")

if __name__ == '__main__':
    unittest.main()